src/rgw/rgw_sal_motr.cc

```diff
@@ -54,6 +57,41 @@ namespace rgw::sal {
 using ::ceph::encode;
 using ::ceph::decode;
 
+class MotrADDBLogger {
+private:
+  uint64_t req_id;
+  bool motr_thread = false;
+public:
+  MotrADDBLogger() {
+    struct m0_thread_tls *tls = m0_thread_tls();
+
+    req_id = (uint64_t)-1;
+    if (tls->tls_addb2_mach == NULL)
+    {
+      m0_addb2_global_thread_enter();
```

Comment:
I'll split my answer into to parts:
1. Why tls is not checked for NULL:
This is because of `m0_thread_tls()` function call - it returns pointer to thread-local-store instance of `struct m0_thread_tls`, and even if tls is NULL, it will allocate and return.
2. Why tls != NULL and tls->tls_addb2_mach == NULL:
Firstly it should be said that RGW is multithreaded application, and only one thread is initialized as Motr thread ( with tls, addb2_mach and all internal stuff). Other threads have to be initialized, or "entered" as a motr global thread. When a thread creates an instance of `class MotrADDBLogger` it doesn't have any tls (and all related variables) initialized, hence any ADDB call will be just ignored - as tls_addb2_mach is not allocated yet. Code above allows thread to operate as a motr thread, and `m0_addb2_global_thread_enter()` allocates addb machine for following logging.